input/evdev_gamepad.c

```diff
@@ -0,0 +1,323 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <linux/input.h>
+#include <linux/input-event-codes.h>
+#include <stdatomic.h>
+#include <stdbool.h>
+#include <sys/ioctl.h>
+#include <poll.h>
+#include <unistd.h>
+#include "common/common.h"
+#include "common/msg.h"
+#include "input.h"
+#include "input/keycodes.h"
+
+struct gamepad {
+    int fd;
+    int left_x_min, left_x_max;
+    int left_y_min, left_y_max;
+    int right_x_min, right_x_max;
+    int right_y_min, right_y_max;
+    int left_trigger, right_trigger;
+    char name[128];
+};
+
+#define MAX_GAMEPADS 4
+
+static atomic_bool gamepad_cancel_wakeup;
+
+static const int button_map[][2] = {
+    { BTN_SOUTH, MP_KEY_GAMEPAD_ACTION_DOWN },
+    { BTN_EAST, MP_KEY_GAMEPAD_ACTION_RIGHT },
+    { BTN_WEST, MP_KEY_GAMEPAD_ACTION_LEFT },
+    { BTN_NORTH,  MP_KEY_GAMEPAD_ACTION_UP },
+    { BTN_SELECT, MP_KEY_GAMEPAD_BACK },
+    { BTN_MODE, MP_KEY_GAMEPAD_MENU },
+    { BTN_START, MP_KEY_GAMEPAD_START },
+    { BTN_THUMBL, MP_KEY_GAMEPAD_LEFT_STICK },
+    { BTN_THUMBR, MP_KEY_GAMEPAD_RIGHT_STICK },
+    { BTN_TL, MP_KEY_GAMEPAD_LEFT_SHOULDER },
+    { BTN_TR, MP_KEY_GAMEPAD_RIGHT_SHOULDER },
+    { BTN_TL2, MP_KEY_GAMEPAD_LEFT_TRIGGER },
+    { BTN_TR2, MP_KEY_GAMEPAD_RIGHT_TRIGGER },
+    { BTN_DPAD_UP, MP_KEY_GAMEPAD_DPAD_UP },
+    { BTN_DPAD_DOWN, MP_KEY_GAMEPAD_DPAD_DOWN },
+    { BTN_DPAD_LEFT, MP_KEY_GAMEPAD_DPAD_LEFT },
+    { BTN_DPAD_RIGHT, MP_KEY_GAMEPAD_DPAD_RIGHT },
+};
+
+static int lookup_button_mp_key(int evdev_key)
+{
+    for (int i = 0; i < MP_ARRAY_SIZE(button_map); i++) {
+        if (button_map[i][0] == evdev_key) {
+            return button_map[i][1];
+        }
+    }
+    return -1;
+}
+
+static int test_device(struct mp_input_src *src, char *name)
+{
+    /* Note that name is always much smaller than 255, but gcc doesn't know that */
+    char path[11 + 255 + 1];
```

Comment:
You could sidestep this by using `openat()` and `fdopendir()`. Then you don't have to perform any path construction.

Reply:
I didn’t need `fdopendir()`, `opendir()` then `dirfd()` were enough to obtain the fd of `/dev/input` to pass to `openat()`, thanks for the suggestion!